src/server.rs

```diff
@@ -15,35 +18,84 @@ pub struct ServerDecodeResult {
 #[rpc(server)]
 trait DecoderRpc {
     #[method(name = "dob_protocol_version")]
-    fn protocol_version(&self) -> String;
+    async fn protocol_version(&self) -> String;
 
     #[method(name = "dob_decode")]
-    fn decode(&self, hexed_spore_id: String) -> Result<String, ErrorCode>;
+    async fn decode(&self, hexed_spore_id: String) -> Result<String, ErrorCode>;
+
+    #[method(name = "dob_batch_decode")]
+    async fn batch_decode(&self, hexed_spore_ids: Vec<String>) -> Result<Vec<String>, ErrorCode>;
 }
 
 pub struct DecoderStandaloneServer {
-    sender: Arc<DecoderCmdSender>,
+    decoder: DOBDecoder,
 }
 
 impl DecoderStandaloneServer {
-    pub fn new(sender: Arc<DecoderCmdSender>) -> Self {
-        Self { sender }
+    pub fn new(decoder: DOBDecoder) -> Self {
+        Self { decoder }
     }
 }
 
+#[async_trait]
 impl DecoderRpcServer for DecoderStandaloneServer {
-    fn protocol_version(&self) -> String {
-        self.sender.protocol_version()
+    async fn protocol_version(&self) -> String {
+        self.decoder.protocol_version()
     }
 
     // decode DNA in particular spore DOB cell
-    fn decode(&self, hexed_spore_id: String) -> Result<String, ErrorCode> {
+    async fn decode(&self, hexed_spore_id: String) -> Result<String, ErrorCode> {
         tracing::info!("decoding spore_id {hexed_spore_id}");
-        let (render_output, dob_content) = self.sender.decode_dna(&hexed_spore_id)?;
+        let spore_id: [u8; 32] = hex::decode(hexed_spore_id)
+            .map_err(|_| Error::HexedSporeIdParseError)?
+            .try_into()
+            .map_err(|_| Error::SporeIdLengthInvalid)?;
+        let mut cache_path = self.decoder.setting().dobs_cache_directory.clone();
+        cache_path.push(format!("{}.dob", hex::encode(&spore_id)));
+        let (render_output, dob_content) = if cache_path.exists() {
+            read_dob_from_cache(cache_path)?
+        } else {
+            let (content, metadata) = self.decoder.fetch_decode_ingredients(spore_id).await?;
+            let render_output = self.decoder.decode_dna(&content, metadata).await?;
+            write_dob_to_cache(&render_output, &content, cache_path)?;
+            (render_output, content)
+        };
         let result = ServerDecodeResult {
             render_output,
             dob_content,
         };
         Ok(serde_json::to_string(&result).unwrap())
     }
+
+    // decode DNA from a set
+    async fn batch_decode(&self, hexed_spore_ids: Vec<String>) -> Result<Vec<String>, ErrorCode> {
```

Comment:
There is a problem with the current return value design: once any of the spore id requests fails, then all the other successful ones can't be returned either.

Perhaps consider returning the result of each item：

```rust
    async fn batch_decode(
        &self,
        hexed_spore_ids: Vec<String>,
    ) -> Result<Vec<Result<String, ErrorCode>>, ErrorCode>
 ```

Reply:
it's a design issue, I consider to remind sender that one of DNAs in the requesting batch is out of service, because if we return others, they may thinks all of these requested DNAs work.

I guess adding a `restrict` flag is a better way to be as a prompt.

Reply:
`Vec<Result<String, ErrorCode>>`, return this type contains Ok or Error and also the content. Every item has a one-to-one result corresponding to the input, like this：

```bash
ethan@T14s-Gen4:~/projects/cell-studio/dob-decoder-standalone-server$ echo '{
    "id": 2,
    "jsonrpc": "2.0",
    "method": "dob_batch_decode",
    "params": [
        [
            "1eba4cbbf9e196a231b0f98951a35eba4cf757f4e1361ba47fb1d75bd2ff2cb6",
            "924dc5eb9f62f9f1741ece4d671185d24eb933edef8b97da6760ef5cb41567b8",
            "9dd9604d44d6640d1533c9f97f89438f17526e645f6c35aa08d8c7d844578580"
        ]
    ]
}' | curl -H 'content-type: application/json' -d @- http://localhost:8090
{"jsonrpc":"2.0","result":[{"Err":1004},{"Err":1007},{"Err":1007}],"id":2}

```